src/libasr/ASR.asdl

```diff
@@ -278,6 +281,7 @@ expr
 
     | BitCast(expr source, expr mold, expr? size, ttype type, expr? value)
     | DerivedRef(expr v, symbol m, ttype type, expr? value)
+    | EnumRef(symbol v, symbol? m, string property, ttype type, expr? value)
```

Comment:
What is the `property` used for?

I am going to setup documentation for ASR, so that we can document (in Markdown) each node and what it does and is meant to be used (for now I was thinking one file per node). We'll also soon have some tests and semantic definition of each ASR node thanks to the work that @rebcabin is doing.

Reply:
We might even use tangledown for this. ;)

Reply:
From reading the C++ code, it seems that `property` is always `value`? I would remove it.

What is the allowed result of `EnumRef` --- what type is it? Is it always the type of the given member?

It seems just the `m` argument is enough to specify. Also, why is it `symbol? m`? What is the meaning if `m==nullptr`?

Reply:
> What is the `property` used for?

It can be either `name` or `value`. For now I only added support for `value` but in follow up PRs I will add support for `name` as well.



> It seems just the `m` argument is enough to specify. Also, why is it `symbol? m`? What is the meaning if `m==nullptr`?

There are two cases,

1. `Color.RED.value` - Here we are accessing `RED` via `Color` class directly. So `EnumRef(Color, RED, value)` is created.
2. `selected_color.value` - Here `selected_color` is an object (from CPython's perspective). So in LPython we create `EnumRef(selected_color, nullptr, value)` (see how `m` is nullptr now).

Reply:
I would distinguish these two cases, it seems they are unrelated.

`Color.RED.value` -> Just use the compile time value of the symbol RED. The symbol `RED` will have a parentsymtab equal to the Enum's symtab, so we can have a function "is_enumerator" that will determine if a given symbol is an enumerator. If we need a special node for this, we could do `EnumeratorValue(RED)`, but given the fact that the implementation just does `RED->m_value` it seems we do not need a node for this?

`a = Color.RED; a.value` -> `(Assign a EnumItem(Color, RED))  (a->m_value)`

Reply:
The tough runtime use case is:
```python
def f(color: Color):
    print(color, color.value)
```
So let's consider:
```
EnumName(symbol v, ttype type, expr? value) # type = string
EnumValue(symbol v, ttype type, expr? value) # type = Enum value type (int, float, str, ...)
```
and the above would be represented by: `EnumName(color, string, nullptr)` and `EnumValue(color, i32, nullptr)`.
The `value` argument is either nullptr (runtime case, such as above), or it contains the compile time result. For example, if you do:
```python
print(Color.RED, Color.RED.value)
```
then you can set `EnumName(color, string, "RED")` and `EnumValue(color, i32, 1)`.

Reply:
It's clear how to implement `EnumValue`, you already did it in this PR. But how would you implement `EnumName` for the runtime case? That is, `EnumName(color, string, "RED")` is clear, but how would you implement `EnumName(color, string, nullptr)`?

Reply:
It seems to implement `EnumName(color, string, nullptr)`, say if Color is of type float, then we have two options:

1. Either we represent Enum as an integer which is an index into two lists: one with names, the other with values.
2. Represent Enum with its type, say f64, then value is immediate, but to get the name, you need to have a runtime hashtable of f64-> str.

Each approach has pros and cons. If we want fast implementation of EnumName, then we should use the first approach (but EnumValue will be slower). If we want fast EnumValue, then the second approach is better, but then the name lookup will be slightly slower.

For iterating, the first approach is better.

I would argue the second approach is probably better, since a very fast lookup of the value is more important. Name lookup as well as iteration is probably secondary. I believe in C and C++, you cannot even lookup the name, nor iterate. But the value retrieval is immediate (of non-consecutive integers).

Reply:
So if we go with the second approach, then I would not even implement EnumName now, we can do that later if needed. I would just implement EnumValue for now.

Reply:
> So if we go with the second approach, then I would not even implement EnumName now, we can do that later if needed. I would just implement EnumValue for now.

Yes that's what I did here. So I think we just need to implement `EnumValue` for now in LLVM and C backend. We will do `EnumName` later.